be/src/storage/compaction_manager.cpp

```diff
@@ -70,34 +70,40 @@ void CompactionManager::_schedule() {
     while (!_stop.load(std::memory_order_consume)) {
         ++_round;
         _wait_to_run();
-        std::shared_ptr<CompactionTask> compaction_task = _try_get_next_compaction_task();
-        if (!compaction_task) {
+        CompactionCandidate compaction_candidate;
+
+        if (!pick_candidate(&compaction_candidate)) {
             std::unique_lock<std::mutex> lk(_mutex);
             _cv.wait_for(lk, 1000ms);
         } else {
-            if (compaction_task->compaction_type() == CompactionType::BASE_COMPACTION) {
-                StarRocksMetrics::instance()->tablet_base_max_compaction_score.set_value(
-                        compaction_task->compaction_score());
+            if (compaction_candidate.type == CompactionType::BASE_COMPACTION) {
+                StarRocksMetrics::instance()->tablet_base_max_compaction_score.set_value(compaction_candidate.score);
             } else {
                 StarRocksMetrics::instance()->tablet_cumulative_max_compaction_score.set_value(
-                        compaction_task->compaction_score());
+                        compaction_candidate.score);
             }
 
-            compaction_task->set_task_id(next_compaction_task_id());
+            auto task_id = next_compaction_task_id();
             LOG(INFO) << "submit task to compaction pool"
-                      << ", task_id:" << compaction_task->task_id()
-                      << ", tablet_id:" << compaction_task->tablet()->tablet_id()
-                      << ", compaction_type:" << starrocks::to_string(compaction_task->compaction_type())
-                      << ", compaction_score:" << compaction_task->compaction_score() << " for round:" << _round
+                      << ", task_id:" << task_id << ", tablet_id:" << compaction_candidate.tablet->tablet_id()
+                      << ", compaction_type:" << starrocks::to_string(compaction_candidate.type)
+                      << ", compaction_score:" << compaction_candidate.score << " for round:" << _round
                       << ", task_queue_size:" << candidates_size();
-            auto st = _compaction_pool->submit_func([compaction_task] { compaction_task->start(); });
+            auto st = _compaction_pool->submit_func([&compaction_candidate, &task_id] {
+                auto compaction_task = compaction_candidate.tablet->create_compaction_task();
+                if (compaction_task == nullptr) {
+                    LOG(WARNING) << "create compaction task failed for tablet:" << compaction_candidate.to_string()
+                                 << ", task_id:" << task_id;
+                    return;
```

Comment:
Why return?

Reply:
create_compaction_task return nullptr means compaction_task is running or disable.